### DISCUSSION TITLE
Bind to unix socket

### DISCUSSION THREAD

**Discussion by doits** [socketry/falcon#152] [Q&A] [answered]
Is it possible to bind to a unix socket? I have this `falcon.rb`:

```rb
#!/usr/bin/env -S falcon host
# frozen_string_literal: true

# Force to always use threads instead of processes/forks
module Falcon
  module Command
    class Host
      def container_class
        Async::Container::Threaded
      end
    end
  end
end

load :rack

hostname = File.basename(__dir__)

rack hostname do
  endpoint Async::HTTP::Endpoint.parse('http://localhost:3000').with(protocol: Async::HTTP::Protocol::HTTP1)
end
```

I tried ...

```rb
endpoint Async::HTTP::Endpoint.parse('unix://./tmp_something').with(protocol: Async::HTTP::Protocol::HTTP1)
```

... but I get:

```
 0.03s    error: Async::Task [oid=0x1568] [ec=0x157c] [pid=62343] [2021-06-16 23:09:27 +0200]
               |   SocketError: getaddrinfo: nodename nor servname provided, or not known
               |   → /.rbenv/versions/2.7.3/lib/ruby/2.7.0/socket.rb:227 in `getaddrinfo'
               |     /.rbenv/versions/2.7.3/lib/ruby/2.7.0/socket.rb:227 in `foreach'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/async-io-1.32.1/lib/async/io/host_endpoint.rb:89 in `each'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/async-http-0.56.3/lib/async/http/endpoint.rb:207 in `each'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/async-io-1.32.1/lib/async/io/endpoint.rb:102 in `bound'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/async-io-1.32.1/lib/async/io/shared_endpoint.rb:31 in `bound'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/falcon-0.39.1/lib/falcon/service/application.rb:68 in `block in start'
               |     /.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/async-1.29.0/lib/async/task.rb:263 in `block in make_fiber'
```

Is there a way to bind to a socket?

**Comment by ioquatix** [comment, marked as answer]
Yes, that's the default: https://github.com/socketry/falcon/blob/4211499316369daba2b6bf949b57b473044c08ed/lib/falcon/environments/application.rb#L51-L63

**Reply by doits** [reply]
Thanks, got it working with the following:

```rb
#!/usr/bin/env -S falcon host
# frozen_string_literal: true

# always force threads instead of processes/forks
Falcon::Command::Host.prepend(
  Module.new do
    def container_class
      Async::Container::Threaded
    end
  end
)

load :rack

hostname = File.basename(__dir__)

rack hostname do
  endpoint ::Falcon::ProxyEndpoint.unix(
    ::File.expand_path('tmp/sockets/falcon', root),
    scheme: 'http',
    protocol: Async::HTTP::Protocol::HTTP1
  )
end

```

**Reply by ioquatix** [reply]
> always force threads instead of processes/forks

What is the reason for this?

**Reply by doits** [reply]
My app works fine in multiple threads but not with multiple forks/processes, because it uses a custom in memory cache store, which additionally might be altered during runtime. With different processes it is not shared and can lead to different results depending on what process answers the request. That's why I am forcing it to threads for now until the app is refactored to not use such kind of store, or do you see another solution?

**Reply by ioquatix** [reply]
I would suggest using https://github.com/socketry/async-redis as a cache backend.

**Reply by uberjay** [reply]
Thanks for this -- I have a relatively small application that falcon is a good fit for (previously I have used EventMachine, but was really not excited about dusting off a huge stack of old async ruby software). Ideally, deployment wouldn't require installing additional services. (iow, no redis).

So far i've just set the application instance to 1 to ensure multiple processes were not started, but hacking falcon to always use a threaded model is better. How would you feel about exposing this as a configuration parameter in the application configuration?
